### PR TITLE
Support arbitrary precision (and various other changes)

### PR DESCRIPTION
**Pull request by AndrolGenhald**
Support arbitrary precision values, rather than losing precision by parsing into a 64-bit float.

Work in progress, I broke #7 while simplifying `format` and `unformat`, but the way it works right now the fix for #7 also seems to cause unexpected behavior, so I think the actual desired behavior for that needs to be thought out a bit more.

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -61,38 +51,30 @@ function format(input, opt = defaults, caller) {
...
   }
-
-  if (opt.allowBlank && isNormalInteger(input)) {
-    input = numbersToCurrency(input, fixed(opt.precision));
-  }
-
-  if (typeof input === 'number') {
-    input = Number(input).toFixed(fixed(opt.precision));
-  } else if (!Number.isNaN(input)) {
-    if (isNormalInteger(input)) {
```
Comment (jonathanpmartins) [inline src/utils.js:75]: Check if a string is an integer

File: src/utils.js [hunk 2]
```diff
@@ -61,38 +51,30 @@ function format(input, opt = defaults, caller) {
   if (opt.debug) console.log('utils format() - caller', caller);
   if (opt.debug) console.log('utils format() - input1', input);
 
-  if (input === null || input === undefined) {
-    input = 0;
+  if (input === null || typeof input === 'undefined') {
+    input = '';
+  } else if (typeof input === 'number') {
+    input = input.toFixed(fixed(opt.precision));
   }
```
Comment (jonathanpmartins) [inline src/utils.js:58]: We could `else` out here and check if a string is an integer

File: src/utils.js [hunk 3]
```diff
@@ -61,38 +51,30 @@ function format(input, opt = defaults, caller) {
   if (opt.debug) console.log('utils format() - caller', caller);
   if (opt.debug) console.log('utils format() - input1', input);
 
-  if (input === null || input === undefined) {
-    input = 0;
+  if (input === null || typeof input === 'undefined') {
+    input = '';
+  } else if (typeof input === 'number') {
+    input = input.toFixed(fixed(opt.precision));
```
Comment (jonathanpmartins) [inline src/utils.js:57]: We could `else` out here and check if a string is an integer

Comment (jonathanpmartins) [inline src/utils.js:57]: We could `else` out here and check if a string is an integer

**Comment by jonathanpmartins** [comment]
Great job! Cleaner is always better. Make the house look pretty! heheh

I need a little more time to take a deep look and think more about it. One thing for sure is that we should support integers or string representation of integers coming in from the database (prop modelValue). I did took a brief look and I believe that integers are correctly covered in your code, but string representations of integers not.

File: src/utils.js (hunk 1 above)
Comment (jonathanpmartins) [inline src/utils.js:75]: Check if a string is an integer. It can be improved, perhaps a regex.

File: src/utils.js (hunk 3 above)
Comment (jonathanpmartins) [inline src/utils.js:57]: We could `else` out here and check if a string is an integer and treat it accordingly.

**Comment by AndrolGenhald** [comment]
Fixed a few things, but #7 behavior is still a little weird. I see a few options:

1. If the input has an integer, it should stay an integer (pasting `"1234"` results in `"1234.00"`)
2. If the input is initialized with an integer it should stay that way, but integers pasted in should be formatted (`"1234"` becomes `"12.34"`)
3. If the input is initialized with type `Number` it will be formatted, but all other cases are formatted (starting with `12.3` results in `"12.30"`, pasting `"12.3"` results in `"1.23"`).

If we want pasting integers to stay as integers, `allowBlank = true` will be a little weird, and I have no idea how to fix it. If the input is blank and the user types `"12"`, it will result in `"10.02"`, because it'll detect the `"1"` as an integer, and the 2 will be appended to the `"1.00"`.

I also added special behavior for `precision === 0`, if the input receives (through starting value or user input) a decimal, it will round it instead of formatting it (eg `"12345.6789"` becomes `"12,346"` instead of `"123,456,789"`).

File: src/utils.js [hunk 4]
```diff
@@ -101,17 +113,10 @@ function format(input, opt = defaults, caller) {
...
-    const currentLength = (`${parts[0]}`).length + (`${parts[1]}`).length;
-    const diff = opt.minimumNumberOfCharacters - currentLength;
-    if (diff > 0) {
-      for (let i = 0; i < diff; i += 1) {
-        parts[0] = `0${parts[0]}`;
-      }
-    }
-  }
+  const decimalLength = parts.length === 2 ? parts[1].length : 0;
+  parts[0].padStart(opt.minimumNumberOfCharacters - decimalLength, '0');
```
Comment (jonathanpmartins) [inline src/utils.js:119]:
```suggestion
  parts[0] = parts[0].padStart(opt.minimumNumberOfCharacters - decimalLength, '0');
```

**Review by jonathanpmartins — COMMENTED** [review]
If you padStart here, I don't think It will be needed above.

https://github.com/AndrolGenhald/v-money3/blob/2557322579826bdf66992d8fbac0aecac7839b85/src/utils.js#L95

File: src/utils.js (hunk 4 above)
Comment (AndrolGenhald) [reply src/utils.js:119]: Yeah, that's a bit weird, removing it causes minimum character tests to fail, I'll investigate that later when I have time.

**Comment by jonathanpmartins** [comment]
> If we want pasting integers to stay as integers, `allowBlank = true` will be a little weird, and I have no idea how to fix it. If the input is blank and the user types `"12"`, it will result in `"10.02"`, because it'll detect the `"1"` as an integer, and the 2 will be appended to the `"1.00"`.

Yes, you caught it.... That weird first type problem. I'm messing with it since the beginning. Tried to fix this before it arrived on the format function.

https://github.com/AndrolGenhald/v-money3/blob/2557322579826bdf66992d8fbac0aecac7839b85/src/directive.js#L73

**Review by jonathanpmartins — APPROVED** [review]

**Review by sha-hin — APPROVED** [review]

**Comment by AndrolGenhald** [comment]
I don't have time at the moment to look into why those puppeteer tests aren't working as expected, I'll just comment them out and let you merge it as is if you think everything else looks good. Hopefully next week I'll have some more time to investigate.

**Comment by jonathanpmartins** [comment]
Cool! No probs...
I'm out of time for now too.
I will probably pull it in a analyze it next week.

**Comment by jonathanpmartins** [comment]
This pull request will probably not be merged. I was working on **BigInt integration** (branch [bigint](https://github.com/jonathanpmartins/v-money3/tree/bigint)) and after some refactoring part of this code wasn't needed anymore. On the other hand part of this code was super helpful and was included by comparing branches (`copy/paste`). Credits to @AndrolGenhald making a really good job analyzing the code and bringing in various edge cases.
Some features and tests not currently included will be included in the final release of this refactoring.

**Comment by AndrolGenhald** [comment]
I was actually thinking about trying to use `BigInt` instead of `string`, but I was already more than half done getting arbitrary precision working with `string`, so I figured I'd save it for later. If you've got it working with `BigInt` that's even better!

**Comment by jonathanpmartins** [comment]
Using `BigInt` will decrease compatibility with old browsers. The problem is that I can only compile the project using `es2020`.
At this point is not just about `BigInt` but everything else that `es2020` support. A lot of new features will not be transpiled down.
